core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -524,6 +523,8 @@ public void onFailure(Exception e) {
                     latch.countDown();
                 }
             }
+            // set this last, once we finished updating all internal state.
+            this.shardRouting = newRouting;
```

Comment:
I think that in the case of a primary this should be published before we countdown the latch that lets the operations that are run under the permit block proceed. At that point we will have published the new primary term but not the new routing entry, and this shard will be used for example in the resync operations. Am I missing something here?

Reply:
good catch. Thanks.